server/src/main/java/org/elasticsearch/action/admin/indices/shards/TransportIndicesShardStoresAction.java

```diff
@@ -135,152 +112,189 @@ protected ClusterBlockException checkBlock(IndicesShardStoresRequest request, Cl
             .indicesBlockedException(ClusterBlockLevel.METADATA_READ, indexNameExpressionResolver.concreteIndexNames(state, request));
     }
 
-    private class AsyncShardStoresInfoFetches {
-        private final DiscoveryNodes nodes;
-        private final RoutingNodes routingNodes;
-        private final Set<Tuple<ShardId, String>> shards;
-        private final ActionListener<IndicesShardStoresResponse> listener;
-        private final RefCountingRunnable refs = new RefCountingRunnable(this::finish);
-        private final Queue<InternalAsyncFetch.Response> fetchResponses;
-
-        AsyncShardStoresInfoFetches(
-            DiscoveryNodes nodes,
-            RoutingNodes routingNodes,
-            Set<Tuple<ShardId, String>> shards,
+    // exposed for tests
+    void listShardStores(
+        TransportNodesListGatewayStartedShards.Request request,
+        ActionListener<TransportNodesListGatewayStartedShards.NodesGatewayStartedShards> listener
+    ) {
+        // async fetch store infos from all the nodes for this one shard
+        // NOTE: instead of fetching shard store info one by one from every node (nShards * nNodes requests)
+        // we could fetch all shard store info from every node once (nNodes requests)
+        // we have to implement a TransportNodesAction instead of using TransportNodesListGatewayStartedShards
+        // for fetching shard stores info, that operates on a list of shards instead of a single shard
+
+        client.executeLocally(TransportNodesListGatewayStartedShards.TYPE, request, listener);
+    }
+
+    private record ShardRequestContext(
+        ShardId shardId,
+        String customDataPath,
+        ActionListener<TransportNodesListGatewayStartedShards.NodesGatewayStartedShards> listener
+    ) {}
+
+    private final class AsyncAction {
+        private final CancellableTask task;
+        private final DiscoveryNode[] nodes;
+        private final String[] concreteIndices;
+        private final RoutingTable routingTable;
+        private final Metadata metadata;
+        private final Map<String, Map<Integer, List<StoreStatus>>> indicesStatuses;
```

Comment:
Why not `Map<ShardId, List<StoreStatus>>` ?

Reply:
Really just because that's what the response wants.